File: Source/Fastore.Core/ColumnHash.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Collections;

namespace Fastore.Core
{
	public class ColumnHash<T>
	{
		public ColumnHash(IComparer<T> comparer = null)
		{
			_comparer = comparer ?? Comparer<T>.Default;
			_values = new BTree<T, ISet<long>>(_comparer);
		}

		private IComparer<T> _comparer;

		private BTree<T, ISet<long>> _values;

		private Dictionary<long, IKeyValueLeaf<T, ISet<long>>> _rows = new Dictionary<long, IKeyValueLeaf<T, ISet<long>>>();

		public Optional<T> GetValue(long rowId)
		{
			IKeyValueLeaf<T, ISet<long>> leaf;
			if (!_rows.TryGetValue(rowId, out leaf))
				return Optional<T>.Null;

			return leaf.GetKey(v => v.Contains(rowId));
		}

		public IEnumerable<KeyValuePair<long, T>> GetRows(bool isForward)
		{
			foreach (var valueEntry in _values.Get(isForward))
				if (isForward)
				{
					foreach (var rowEntry in valueEntry.Value)
						yield return new KeyValuePair<long, T>(rowEntry, valueEntry.Key);
				}
				else
				{
					foreach (var rowEntry in valueEntry.Value.Reverse())
						yield return new KeyValuePair<long, T>(rowEntry, valueEntry.Key);
				}
		}

		public IEnumerable<KeyValuePair<long, T>> GetRows(bool isForward, Optional<T> start)
		{
			foreach (var valueEntry in _values.Get(isForward, start))
				if (isForward)
				{
					foreach (var rowEntry in valueEntry.Value)
						yield return new KeyValuePair<long, T>(rowEntry, valueEntry.Key);
				}
				else
				{
					foreach (var rowEntry in valueEntry.Value.Reverse())
						yield return new KeyValuePair<long, T>(rowEntry, valueEntry.Key);
				}
		}

		public IEnumerable<KeyValuePair<long, T>> GetRows(bool isForward, Optional<T> start, Optional<T> end, int? limit)
		{
			var items = _values.Get(isForward, start, end);
			foreach (var valueEntry in limit.HasValue ? items.Take(limit.Value) : items)
				if (isForward)
				{
					foreach (var rowEntry in valueEntry.Value)
						yield return new KeyValuePair<long, T>(rowEntry, valueEntry.Key);
				}
				else
				{
					foreach (var rowEntry in valueEntry.Value.Reverse())
						yield return new KeyValuePair<long, T>(rowEntry, valueEntry.Key);
				}
		}

		public bool Insert(T value, long rowId)
		{
			// TODO: avoid constructing row hashset until needed

			// Create a new row bucket in case there isn't only for the given value  
			ISet<long> newRows = new HashSet<long>();

			// Attempt to insert the row bucket
			IKeyValueLeaf<T, ISet<long>> valueLeaf;
			var existingRows = _values.Insert(value, newRows, out valueLeaf);
			// If already existing, use it
			if (existingRows != Optional<ISet<long>>.Null)
				newRows = existingRows.Value;

			// Add the row into the row bucket
			if (newRows.Add(rowId))
			{
				// If a row was added, add the row leaf to the hash table
				_rows.Add(rowId, valueLeaf);

				return true;
			}
			else
				return false;
		}
	}
}

File: Source/BtreeCPP/ColumnHash.cpp
#include "ColumnHash.h"
#include <hash_set>

ColumnHash::ColumnHash(int(*compare)(void* left, void* right), wstring(*tostring)(void*))
{
	_compare = compare;

	_values = new BTree(128,128, _compare, tostring);

	_values->Observer = (IObserver*)this;
	_rows = new hash_map<long, Leaf*>();
}

void* ColumnHash::GetValue(long rowID)
{
	hash_map <long, Leaf*> :: const_iterator hash_mapIterator;
	hash_mapIterator = _rows->find(rowID);
	
	if(hash_mapIterator != _rows->end())
	{
		Leaf* leaf = hash_mapIterator->second;
		return leaf->GetKey(
			[rowID](void* hash) -> bool
			{
				hash_set<long>* newrows = (hash_set<long>*)hash;
				return newrows->find(rowID) != newrows->end();
			});
	}
	else
	{
		return NULL;
	}
}

typedef pair <long,Leaf*> LongLeafPair;
bool ColumnHash::Insert(long rowID, void* value)
{
	Leaf* valueLeaf;
	hash_set<long>* newrows = new hash_set<long>();
	void* existing = _values->Insert(value, newrows, valueLeaf);
	if(existing != NULL)
		newrows = (hash_set<long>*)existing;

	if(newrows->insert(rowID).second)
	{
		_rows->insert(LongLeafPair (rowID, valueLeaf));
		return true;
	}
	else
		return false;

}

void ColumnHash::ValuesMoved(void* lp)
{
	Leaf* leaf = (Leaf *)lp;
	for(int i = 0; i < leaf->Count; i++)
	{
		hash_set<long>* current = (hash_set<long>*)leaf->_values[i];
		hash_set<long> :: const_iterator hash_setIterator;
		hash_setIterator = current->begin();
		while(hash_setIterator != current->end())
		{
			_rows->find(*hash_setIterator)->second = leaf;
			hash_setIterator++;
		}
	}
}